File: Implementations/PrimeCheck.hs
import Text.Printf

(^!) :: Num a => a -> Int -> a
(^!) x n = x^n

squareRoot :: Integer -> Integer
squareRoot 0 = 0
squareRoot 1 = 1
squareRoot n =
   let twopows = iterate (^!2) 2
       (lowerRoot, lowerN) =
          last $ takeWhile ((n>=) . snd) $ zip (1:twopows) twopows
       newtonStep x = div (x + div n x) 2
       iters = iterate newtonStep (squareRoot (div n lowerN) * lowerRoot)
       isRoot r  =  r^!2 <= n && n < (r+1)^!2
  in  head $ dropWhile (not . isRoot) iters
  
loop i num = do
    if (num `mod` i == 0) then
          printf "%d is not prime: %d*%d\n" num i (num `div` i)
    else 
        if i < (squareRoot num) then 
            loop (i + 1) num
        else do
            printf "%d is prime\n" num
    
    
main = do
    putStrLn "Enter a number"
    numStr <- getLine
    let num = (read numStr  :: Integer)
    loop 2 num

File: Implementations/PrimeCheck.go
package main

import (
    "fmt"
    "math/big"
)

func main() {
    num := big.NewInt(0)
    sqrt := big.NewInt(0)
    tmp := big.NewInt(0)
    fmt.Println("Enter a number")
    str := ""
    fmt.Scan(&str)
    num.SetString(str, 10)
    for  i := big.NewInt(2); i.Cmp(sqrt.Add(sqrt.Sqrt(num), big.NewInt(1))) == -1; i.Add(i, big.NewInt(1)) {
        tmp.Mod(num, i)
        if  tmp.Cmp(big.NewInt(0)) == 0 {
            fmt.Print(num, " is not prime: ", i, "*", tmp.Div(num, i), "\n")
            return
        }
    }
    fmt.Println(num, "is prime")
    return
}
